src/Utilities/DoubleDicts.jl

```diff
@@ -299,6 +299,59 @@ function Base.keys(d::AbstractDoubleDictInner{F,S}) where {F,S}
     return MOI.ConstraintIndex{F,S}.(keys(d.dict))
 end
 
+"""
+    outer_keys(d::AbstractDoubleDict)
+
+Return an iterator over the outer keys of the `AbstractDoubleDict` `d`.
+Each outer key is a `Tuple{Type,Type}` so that a double loop can
+be easily used:
+```julia
+for (F, S) in DoubleDicts.outer_keys(dict)
+    for (k, v) in dict[F, S]
+        # ...
+    end
+end
+```
+
+For performance, it is recommended that the inner loop lies in a separate
+function to gurantee type-stability.
+Some outer keys `(F, S)` might lead to an empty `dict[F, S]`.
+If you want only nonempty `dict[F, S]`, use [`outer_keys_vector`](@ref).
+"""
+outer_keys(d::AbstractDoubleDict) = keys(d.dict)
```

Comment:
For consistency, this one could also have the `only_nonempty` keyword and return a `Base.Iterators.Filter` in that case. The return type would then be a `Union` though

Reply:
how about:

Keep the current `outer_keys` and add:

```julia
function nonempty_outer_keys(d::AbstractDoubleDict)
    # using Base.Iterators.Filter
end
```

?

Reply:
Yes, I would prefer that. It's easy for the user to call `collect` on top of either functions if a vector is needed